### PR TITLE
feat(ui): add generic networking fields

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Add a number of fields for networking forms.
- Update the add machine form to use the mac address
field.

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to the machine list and open the add machine form.
- Check that the mac address fields works as before.

## Fixes

Fixes: #2185.

### CODE REVIEW

File: ui/src/app/base/components/FabricField/FabricField.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  if (!fabricsLoaded) {
+    return <Spinner />;
+  }
+
+  return (
+    <FormikField
+      component={Select}
+      label="Fabric"
+      name="fabric"
+      options={fabrics.map((fabric) => ({
```
Comment (Caleb-Ellis) [inline ui/src/app/base/components/FabricField/FabricField.tsx:31]: In most selects I've included a disabled default option that just says "Select foo" - it was a carry over from the angular version but we should at least keep them consistent

File: ui/src/app/base/components/SubnetField/SubnetField.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  filterFunction?: (subnet: Subnet) => boolean;
+} & Partial<FormikFieldProps>;
+
+export const SubnetField = ({
+  defaultOption,
+  filterFunction,
+  ...props
+}: Props): JSX.Element => {
+  const dispatch = useDispatch();
+  let subnets: Subnet[] = useSelector(subnetSelectors.all);
```
Comment (Caleb-Ellis) [inline ui/src/app/base/components/SubnetField/SubnetField.tsx:24]: Does this need to be typed?

**Review by Caleb-Ellis — CHANGES_REQUESTED** [review]
Mostly looks good, my only change request is that `name` be required on all the formik fields. The rest are mainly just to keep things consistent with the other generic fields, but that can come later maybe when we've decided how everything *should* work.

File: ui/src/app/base/components/SubnetField/SubnetField.tsx (hunk 2 above)
Comment (huwshimi) [reply ui/src/app/base/components/SubnetField/SubnetField.tsx:24]: For some reason these are coming back as `unknown` so when the `.map` happens below it says `Property 'map' does not exist on type 'unknown'`.

File: ui/src/app/base/components/FabricField/FabricField.tsx (hunk 1 above)
Comment (huwshimi) [reply ui/src/app/base/components/FabricField/FabricField.tsx:31]: OK, I'll add them.

File: ui/src/app/base/components/FabricField/FabricField.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import { useDispatch, useSelector } from "react-redux";
+
+import FormikField from "app/base/components/FormikField";
+import type { Props as FormikFieldProps } from "app/base/components/FormikField/FormikField";
+import { actions as fabricActions } from "app/store/fabric";
+import fabricSelectors from "app/store/fabric/selectors";
+
+type Props = Partial<FormikFieldProps>;
+
+export const FabricField = ({ ...props }: Props): JSX.Element => {
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply ui/src/app/base/components/FabricField/FabricField.tsx:13]: Sure, I'll make those changes.

File: ui/src/app/base/components/SubnetField/SubnetField.tsx (hunk 2 above)
Comment (Caleb-Ellis) [reply ui/src/app/base/components/SubnetField/SubnetField.tsx:24]: Right, I guess it could be because it's declared as a `let`, which could change to be anything.

Comment (huwshimi) [reply ui/src/app/base/components/SubnetField/SubnetField.tsx:24]: Even with `let` you'd need to explicitly cast it to another type if you wanted to redeclare it with something else. I think it's more likely our selector util doesn't maintain the types for some reason.

File: ui/src/app/machines/components/ActionFormWrapper/TestForm/TestForm.test.tsx [hunk 4]
```diff
@@ -157,17 +157,20 @@ describe("TestForm", () => {
   it("prepopulates scripts of a given hardwareType", () => {
     const state = { ...initialState };
     const networkScript = scriptsFactory({
+      name: "test1",
```
Comment (huwshimi) [inline ui/src/app/machines/components/ActionFormWrapper/TestForm/TestForm.test.tsx:160]: Drive-by fix for duplicate key warnings in the tests.

**Comment by huwshimi** [comment]
@Caleb-Ellis this is ready for another look.

**Review by Caleb-Ellis — APPROVED** [review]
LGTM!
